presto-raptor/src/main/java/com/facebook/presto/raptor/storage/BucketBalancer.java

```diff
@@ -0,0 +1,285 @@
+ *
+ * - For a given distribution, each node should be allocated the same number of buckets. This enhances parallelism, and therefore query performance.
+ * - The total disk utilization across the cluster should be balanced. This ensures that total cluster storage capacity is maximized. Simply allocating the same number of buckets
+ *   to every node may not achieve this, as bucket sizes may vary dramatically across distributions.
+ *
+ * This prioritizes query performance over total cluster storage capacity.
+ */
+public class BucketBalancer
+{
+    private static final Logger log = Logger.get(BucketBalancer.class);
+
+    private final NodeSupplier nodeSupplier;
+    private final ShardManager shardManager;
+    private final Duration interval;
+    private final boolean backupAvailable;
+    private final boolean coordinator;
+    private final ScheduledExecutorService executor;
+
+    private final AtomicBoolean started = new AtomicBoolean();
+
+    private final CounterStat bucketsBalanced = new CounterStat();
+    private final CounterStat jobErrors = new CounterStat();
+
+    @Inject
+    public BucketBalancer(
+            NodeManager nodeManager,
+            NodeSupplier nodeSupplier,
+            ShardManager shardManager,
+            StorageManagerConfig config,
+            BackupService backupService,
+            RaptorConnectorId connectorId)
+    {
+        this(nodeSupplier,
+                shardManager,
+                config.getBalancerInterval(),
+                backupService.isBackupAvailable(),
+                nodeManager.getCurrentNode().isCoordinator(),
+                connectorId.toString());
+    }
+
+    public BucketBalancer(
+            NodeSupplier nodeSupplier,
+            ShardManager shardManager,
+            Duration interval,
+            boolean backupAvailable,
+            boolean coordinator,
+            String connectorId)
+    {
+        this.nodeSupplier = requireNonNull(nodeSupplier, "nodeSupplier is null");
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.interval = requireNonNull(interval, "interval is null");
+        this.backupAvailable = backupAvailable;
+        this.coordinator = coordinator;
+        this.executor = newScheduledThreadPool(1, daemonThreadsNamed("shard-balancer-" + connectorId));
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (backupAvailable && coordinator && !started.getAndSet(true)) {
+            startJob();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        executor.shutdownNow();
+    }
+
+    @Managed
+    @Nested
+    public CounterStat getBucketsBalanced()
+    {
+        return bucketsBalanced;
+    }
+
+    @Managed
+    @Nested
+    public CounterStat getJobErrors()
+    {
+        return jobErrors;
+    }
+
+    private void startJob()
+    {
+        executor.scheduleWithFixedDelay(() -> {
```

Comment:
Agreed, but doing this at shorter intervals reduces the volume of imbalance anyway, so this might work. Synchronizing execution properly is a separate project, though.